kernel/src/hil/flash.rs

```diff
@@ -4,141 +4,150 @@
 //!         FlashUser {
 //!             driver: driver,
 //!             buffer: TakeCell::new(buffer),
 //!         }
 //!     }
 //! }
 //!
-//! impl<'a, F: hil::flash::Flash> hil::flash::Client<F> for FlashUser<'a, F> {
-//!     fn read_complete(&self, buffer: &'static mut F::Page, error: hil::flash::Error) {}
-//!     fn write_complete(&self, buffer: &'static mut F::Page, error: hil::flash::Error) { }
-//!     fn erase_complete(&self, error: hil::flash::Error) {}
+//! impl<'a, F: flash::Flash<W, E>, const W: usize, const E: usize> flash::Client<W, E> for FlashUser<'a, F, W, E> {
+//!     fn read_complete(&self, read_buffer: &'static mut [u8], ret: Result<(), ReturnCode>) {}
+//!     fn write_complete(&self, write_buffer: &'static mut [u8; W], ret: Result<(), ReturnCode>) { }
+//!     fn erase_complete(&self, ret: Result<(), ReturnCode>) {}
 //! }
 //! ```
 
 use crate::returncode::ReturnCode;
 
-/// Flash errors returned in the callbacks.
-#[derive(Copy, Clone, Debug, Eq, PartialEq)]
-pub enum Error {
-    /// Success.
-    CommandComplete,
-
-    /// An error occurred during the flash operation.
-    FlashError,
-}
-
 pub trait HasClient<'a, C> {
     /// Set the client for this flash peripheral. The client will be called
     /// when operations complete.
     fn set_client(&'a self, client: &'a C);
 }
 
-/// A page of writable persistent flash memory.
-pub trait Flash {
-    /// Type of a single flash page for the given implementation.
-    type Page: AsMut<[u8]> + Default;
-
-    /// Read a page of flash into the buffer.
-    fn read_page(
+/// A page of writeable persistent flash memory.
+///
+/// `W`: Should be the minimum number of bytes that can be written
+///      in an operation.
+/// `E`: Should be the minimum number of bytes that can be erased
+///      in an operation.
+pub trait Flash<const W: usize, const E: usize> {
+    /// Read data from flash into a buffer.
+    ///
+    /// This function will read data stored in flash at `address` and
+    /// `length` into the buffer `buf`.
+    ///
+    /// Not all implementations support this, in that case `read_page`
+    /// should be used instead.
+    ///
+    /// On success returns nothing
+    /// On failure returns a `ReturnCode` and the buffer passed in.
+    fn read(
         &self,
-        page_number: usize,
-        buf: &'static mut Self::Page,
-    ) -> Result<(), (ReturnCode, &'static mut Self::Page)>;
+        address: usize,
+        length: usize,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ReturnCode, &'static mut [u8])>;
 
-    /// Write a page of flash from the buffer.
-    fn write_page(
+    /// Write data from a buffer to flash.
+    ///
+    /// This function will write the buffer `buf` to the `address` specified
+    /// in flash.
+    ///
+    /// `address` should be aligned to `W`.
+    ///
+    /// This function will not erase the page first. The user of this function
+    /// must ensure that a page is erased before writing.
```

Comment:
Should we more precisely encode what it means to erase versus write in this HIL? It's probably meaningful to encode that writes allow 1->0 but not 0->1 explicitly

Reply:
I think users of flash would know that, but I'll add something,